server/src/main/java/org/elasticsearch/index/codec/vectors/ES815BitFlatVectorsFormat.java

```diff
@@ -0,0 +1,139 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.codec.vectors;
+
+import org.apache.lucene.codecs.hnsw.FlatVectorsFormat;
+import org.apache.lucene.codecs.hnsw.FlatVectorsReader;
+import org.apache.lucene.codecs.hnsw.FlatVectorsScorer;
+import org.apache.lucene.codecs.hnsw.FlatVectorsWriter;
+import org.apache.lucene.codecs.lucene99.Lucene99FlatVectorsFormat;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.util.VectorUtil;
+import org.apache.lucene.util.hnsw.RandomAccessVectorValues;
+import org.apache.lucene.util.hnsw.RandomVectorScorer;
+import org.apache.lucene.util.hnsw.RandomVectorScorerSupplier;
+import org.apache.lucene.util.quantization.RandomAccessQuantizedByteVectorValues;
+
+import java.io.IOException;
+
+class ES815BitFlatVectorsFormat extends FlatVectorsFormat {
+
+    private final FlatVectorsFormat delegate = new Lucene99FlatVectorsFormat(FlatBitVectorScorer.INSTANCE);
+
+    @Override
+    public FlatVectorsWriter fieldsWriter(SegmentWriteState segmentWriteState) throws IOException {
+        return delegate.fieldsWriter(segmentWriteState);
+    }
+
+    @Override
+    public FlatVectorsReader fieldsReader(SegmentReadState segmentReadState) throws IOException {
+        return delegate.fieldsReader(segmentReadState);
+    }
+
+    static class FlatBitVectorScorer implements FlatVectorsScorer {
+
+        static final FlatBitVectorScorer INSTANCE = new FlatBitVectorScorer();
+
+        static void checkDimensions(int queryLen, int fieldLen) {
+            if (queryLen != fieldLen) {
+                throw new IllegalArgumentException("vector query dimension: " + queryLen + " differs from field dimension: " + fieldLen);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return super.toString();
+        }
+
+        @Override
+        public RandomVectorScorerSupplier getRandomVectorScorerSupplier(
+            VectorSimilarityFunction vectorSimilarityFunction,
+            RandomAccessVectorValues randomAccessVectorValues
+        ) throws IOException {
+            if (randomAccessVectorValues instanceof RandomAccessVectorValues.Bytes randomAccessVectorValuesBytes) {
+                assert randomAccessVectorValues instanceof RandomAccessQuantizedByteVectorValues == false;
+                return switch (vectorSimilarityFunction) {
+                    case DOT_PRODUCT, MAXIMUM_INNER_PRODUCT, COSINE, EUCLIDEAN -> new HammingScorerSupplier(randomAccessVectorValuesBytes);
+                };
+            }
+            throw new IllegalArgumentException("Unsupported vector type");
```

Comment:
should we also add "Unsupported vector type or similarity function", or we don't expect changes to `VectorSimilarityFunction`?

Reply:
I can add that to the error, but if the function does ever get updated, we would fail to compile, so we would likely have to add some handling for it anyways.